### PR TITLE
Extend Table:MultiV to Six Independent Variables

### PR DESCRIPTION
**Pull request by Myoldmopar**
Pull request overview
---------------------
This task is pretty straightforward: allow 6 independent variables in the multivariable table lookup.  Right now I'm getting it going to allow a separate project to make use of it, but it will eventually go into the main code in **some** form (see @jasondegraw's work to refactor curve manager...).  This doesn't really necessitate a design doc, but if anyone wants one I'll whip one up.  Unit tests and integration tests that exercise this will also accompany any final merge into develop.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [X] ExpandObjects changes? I am not sure if expand objects writes tables, but marking this for now just in case.
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

File: idd/Energy+.idd.in [hunk 1]
```diff
@@ -85419,11 +85419,17 @@ Table:MultiVariableLookup,
   N12, \field Maximum Value of X5
        \type real
        \unitsBasedOnField A12
-  N13, \field Minimum Table Output
+  N13, \field Minimum Value of X6
```
Comment (Myoldmopar) [inline idd/Energy+.idd.in:85422]: New min and max for X6, increment N-count by 2

File: idd/Energy+.idd.in [hunk 2]
```diff
@@ -85472,177 +85478,204 @@ Table:MultiVariableLookup,
        \key Power
        \key Distance
        \default Dimensionless
-  A13, \field Output Unit Type
+  A13, \field Input Unit Type for X6
```
Comment (Myoldmopar) [inline idd/Energy+.idd.in:85481]: New unit type for X6, increment A-count by 1

File: idd/Energy+.idd.in [hunk 3]
```diff
@@ -85472,177 +85478,204 @@ Table:MultiVariableLookup,
...
-  N2960,N2961,N2962,N2963,N2964,N2965,N2966,N2967,N2968,N2969,N2970,N2971,N2972,N2973,N2974,N2975,N2976,N2977,N2978,N2979, \note fields as indicated
-  N2980,N2981,N2982,N2983,N2984,N2985,N2986,N2987,N2988,N2989,N2990,N2991,N2992,N2993,N2994,N2995,N2996,N2997,N2998,N2999, \note fields as indicated
-  N3000,N3001,N3002,N3003,N3004,N3005,N3006,N3007,N3008,N3009,N3010,N3011,N3012,N3013,N3014,N3015,N3016,N3017,N3018,N3019, \note fields as indicated
-  N3020,N3021,N3022,N3023,N3024,N3025,N3026,N3027,N3028,N3029,N3030,N3031,N3032,N3033,N3034,N3035,N3036,N3037,N3038,N3039, \note fields as indicated
-  N3040,N3041,N3042,N3043,N3044,N3045,N3046,N3047,N3048,N3049,N3050,N3051,N3052,N3053,N3054,N3055,N3056,N3057,N3058,N3059, \note fields as indicated
-  N3060,N3061,N3062,N3063,N3064,N3065,N3066,N3067,N3068,N3069,N3070,N3071,N3072,N3073,N3074,N3075,N3076,N3077,N3078,N3079, \note fields as indicated
-  N3080,N3081,N3082,N3083,N3084,N3085,N3086,N3087,N3088,N3089,N3090,N3091,N3092,N3093,N3094,N3095,N3096,N3097,N3098,N3099; \note fields as indicated
+N22,N23,N24,N25,N26,N27,N28,N29,N30,N31,N32,N33,N34,N35,N36,N37,N38,N39,N40,N41,N42,N43,N44,N45,N46,  \note fields as indicated
```
Comment (Myoldmopar) [inline idd/Energy+.idd.in:85507]: Increment all N values by 2.  I did this by hand, so there may be typos.

File: scripts/dev/make_multiv_table_fields.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env python
```
Comment (Myoldmopar) [inline scripts/dev/make_multiv_table_fields.py:1]: Here is a small snippet that could be useful if we ever want to change field numbers for that object again.

File: src/EnergyPlus/CurveManager.cc [hunk 5]
```diff
@@ -285,7 +285,8 @@ namespace CurveManager {
 		Optional< Real64 const > Var2, // 2nd independent variable
 		Optional< Real64 const > Var3, // 3rd independent variable
 		Optional< Real64 const > Var4, // 4th independent variable
-		Optional< Real64 const > Var5 // 5th independent variable
+		Optional< Real64 const > Var5, // 5th independent variable
+		Optional< Real64 const > Var6  // 6th independent variable
```
Comment (Myoldmopar) [inline src/EnergyPlus/CurveManager.cc:289]: Add a new optional 6th argument to the CurveValue evaluation function

File: src/EnergyPlus/CurveManager.cc [hunk 6]
```diff
@@ -458,7 +460,7 @@ namespace CurveManager {
 		int NumTables; // Total tables in the input file
 		int CurveIndex; // do loop index
 		int CurveNum; // current curve number
-		Array1D_string Alphas( 13 ); // Alpha items for object
+		Array1D_string Alphas( 14 ); // Alpha items for object
```
Comment (Myoldmopar) [inline src/EnergyPlus/CurveManager.cc:463]: Alphas needed to be extended by 1 for the unit type field in the multivariable lookup objects.  I didn't increment Numbers up from 10000 to 10002, but I can.  But I probably won't.

File: src/EnergyPlus/CurveManager.cc [hunk 7]
```diff
@@ -1908,7 +1910,7 @@ namespace CurveManager {
 			if ( NumAlphas >= 5 ) {
 				if ( ! IsCurveOutputTypeValid( Alphas( 5 ) ) ) {
 					ShowSevereError( "GetCurveInput: For " + CurrentModuleObject + ": " + Alphas( 1 ) );
-					ShowContinueError( cAlphaFieldNames( 5 ) + " [" + Alphas( 5 ) + "] is invlaid" );
+					ShowContinueError( cAlphaFieldNames( 5 ) + " [" + Alphas( 5 ) + "] is invalid" );
```
Comment (Myoldmopar) [inline src/EnergyPlus/CurveManager.cc:1913]: Pronouncing `invlaid` makes me think of the word "to blave"

image

File: src/EnergyPlus/CurveManager.cc [hunk 8]
```diff
@@ -2038,9 +2040,9 @@ namespace CurveManager {
 			TableLookup( TableNum ).NumIndependentVars = 1;
 			TableLookup( TableNum ).NumX1Vars = size( PerfCurveTableData( TableNum ).X1 );
 			TableLookup( TableNum ).X1Var.allocate( TableLookup( TableNum ).NumX1Vars );
-			TableLookup( TableNum ).TableLookupZData.allocate( 1, 1, 1, 1, size( PerfCurveTableData( TableNum ).Y ) );
+			TableLookup( TableNum ).TableLookupZData.allocate( 1, 1, 1, 1, 1, size( PerfCurveTableData( TableNum ).Y ) );
```
Comment (Myoldmopar) [inline src/EnergyPlus/CurveManager.cc:2043]: ZData now needs an extra dimension, so wherever it's not used (5 dimensions or less), just include scalar dimensionality like already done for lower rank matrices.

File: src/EnergyPlus/CurveManager.cc [hunk 9]
```diff
@@ -2924,7 +2926,18 @@ namespace CurveManager {
 				PerfCurve( CurveNum ).Var5Max = Numbers( 12 );
 				PerfCurve( CurveNum ).Var5MaxPresent = true;
 			}
-
+			if( lNumericFieldBlanks( 13 ) ) {
```
Comment (Myoldmopar) [inline src/EnergyPlus/CurveManager.cc:2929]: Process the new min/max input fields.

File: src/EnergyPlus/CurveManager.cc [hunk 10]
```diff
@@ -3385,20 +3412,21 @@ namespace CurveManager {
 				gio::write( OutputFileInits, Format_130 );
 				gio::write( OutputFileInits, Format_150 ) << CurrentModuleObject << Alphas( 1 );
 			}
-			NumIVars = Numbers( 15 );
-			if ( NumIVars > 5 || NumIVars < 1 ) {
+			NumIVars = Numbers( 17 );
```
Comment (Myoldmopar) [inline src/EnergyPlus/CurveManager.cc:3415]: Grab the number of independent variables from two N lower, and adjust the `BaseOffset` a few lines below up by 2 also.

File: src/EnergyPlus/CurveManager.cc [hunk 11]
```diff
@@ -3448,6 +3479,7 @@ namespace CurveManager {
 					for ( I = 1; I <= TableLookup( TableNum ).NumX1Vars; ++I ) {
 						TableLookup( TableNum ).X1Var( I ) = Numbers( BaseOffset + NumIVars + I );
 					}
+					std::vector< Real64 > x1values( TableLookup( TableNum ).X1Var.begin(), TableLookup( TableNum ).X1Var.end() );
```
Comment (Myoldmopar) [inline src/EnergyPlus/CurveManager.cc:3482]: This is a debug line that I will remove.

File: src/EnergyPlus/CurveManager.cc [hunk 12]
```diff
@@ -3670,18 +3702,68 @@ namespace CurveManager {
 								}
 							}
 
-							TableLookup( TableNum ).TableLookupZData.allocate( TableLookup( TableNum ).NumX5Vars, TableLookup( TableNum ).NumX4Vars, TableLookup( TableNum ).NumX3Vars, TableLookup( TableNum ).NumX2Vars, TableLookup( TableNum ).NumX1Vars );
+							if ( NumIVars > 5 ) {  // something is wrong with this structure here...
```
Comment (Myoldmopar) [inline src/EnergyPlus/CurveManager.cc:3705]: This structure is no longer wrong, my comment is out of date.

File: src/EnergyPlus/CurveManager.cc [hunk 13]
```diff
@@ -4780,6 +4944,53 @@ Label999: ;
...
+			}
+			// perform 2-D interpolation of 2D array to find final value
+			TableValue = DLAG( V5, V6, VALSV5, VALSV6, TWODVALS, NV5, NV6, NUMPT, IEXTV5, IEXTV6 );
+			TWODVALS.deallocate();
+			FOURDVALS.deallocate();
+			VALSX.deallocate();
+			VALSY.deallocate();
+			VALSV3.deallocate();
+			VALSV4.deallocate();
+			VALSV5.deallocate();
```
Comment (Myoldmopar) [inline src/EnergyPlus/CurveManager.cc:4993]: I need to add VALSV6 deallocation here to match the rest.

**Review by Myoldmopar — COMMENTED** [review]
Even from my own self-review, I found three items to address:
 - Remove debug variable
 - Remove outdated comment
 - Add deallocation of 6th independent variable temp array

**Comment by Myoldmopar** [comment]
Good news that the two modified files have expected audit/table diffs...interesting that Windows didn't pick them up though.

**Comment by nrel-bot-2** [comment]
@Myoldmopar @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@Myoldmopar @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@Myoldmopar @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@Myoldmopar @lgentile it has been 14 days since this pull request was last updated.

File: idd/Energy+.idd.in (hunk 3 above)
Comment (rraustad) [reply idd/Energy+.idd.in:85507]: Good grief. I was hoping you deleted these and did some tricks. Then I scrolled down!

File: src/EnergyPlus/CurveManager.cc (hunk 6 above)
Comment (rraustad) [reply src/EnergyPlus/CurveManager.cc:463]: The last N I see above is N3100, your probably safe.

File: src/EnergyPlus/CurveManager.cc (hunk 13 above)
Comment (rraustad) [reply src/EnergyPlus/CurveManager.cc:4993]: How could you have missed that?

**Comment by rraustad** [comment]
Once through review, nothing stands out as wrong (from the code I could see in this view). I hope there is a test file.

**Comment by rraustad** [comment]
I made a second pass, no comments. If a test file works and LookupTable still works, then there is nothing holding this up. VALSV6.deallocate still needs to be added.

**Comment by rraustad** [comment]
I want to drop this in. Can you acknowledge a 6-var file is read in and processed correctly? as is a 5-var file?

**Comment by Myoldmopar** [comment]
I want to also, but we shouldn't right now.  The only feature needing it at the moment is the hybrid air conditioner, and that branch already has the six independent variable table in it, with some modifications.  This branch should actually just go away probably.  I'll close this PR but leave the branch for now.

When the hybrid air conditioner model branch goes in, it will bring along a proper 6 independent variable table.

**Pull request closed without merging**